### ISSUE TITLE
substantial overhead in tf.GradientTape.jacobian?

### ISSUE THREAD

**Issue by yhidaka** [tensorflow/tensorflow#48314]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Scientific Linux 7.9 (Nitrogen)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): conda defaults channel (using "tensorflow-gpu", even though only CPUs are being used at the moment)
- TensorFlow version (use command below): unknown 2.4.1
- Python version: 3.8.8
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

There seems to be a substantial overhead in using the tape.jacobian(), compared to running a for-loop with tape.gradient().

** When "n_div = 3" in the code below:

Forward pass took 0.015 [s]
jacobian took 3.562 [s]
3 gradient calls took 0.055 [s]

** When "n_div = 1001" in the code below:

Forward pass took 0.014 [s]
jacobian took 3.549 [s]
1001 gradient calls took 18.359 [s]

**Describe the expected behavior**

The direct use of a for-loop with tape.gradient() implies that each backpropagation takes almost the same time as the forward pass, and the total running time scales roughly linearly with "n_div". This agrees with my expectation.

I was expecting tape.jacobian() to take, at most, the running time of the tape.gradient() with a for-loop in the worst case scenario of not utilizing any parallelism, and at best, almost the same time as one backpropagation. However, the jacobian's running time appears dominated by some overhead actions.

The server I tested the minimal code below has no GPU installed. Eventually I want to run a more complicated version of this on a cluster node with a GPU with the tape.jacobian() taking only one backpropagation time for "n_div" ~ 100.

**Standalone code to reproduce the issue**

```
import time

import numpy as np
import tensorflow as tf

def simple_func(x0, dp0):

    x1 = x0 / (1 + dp0) * 1.0001

    return x1

def scalar_example(n_repeat, dp0=0.0, show_print=True):

    x0 = tf.constant(0.0, name='x0')
    dp0 = tf.constant(dp0, name='dp0')

    t0 = time.time()
    x1, dp1 = x0, dp0
    with tf.GradientTape(persistent=False) as tape:
        tape.watch(x0)
        tape.watch(dp0)

        for _ in range(n_repeat):
            x1 = simple_func(x1, dp1)
    if show_print:
        print(f'Forward pass took {time.time()-t0:.3f} [s]')

    t0 = time.time()
    m11 = tape.gradient(x1, x0)
    if show_print:
        print(f'Gradient calc for m11 took {time.time()-t0:.3f} [s]')

        print('Finished')

    return m11

def vector_example(n_repeat, n_div):

    x0 = tf.constant(0.0, name='x0')
    dp0 = tf.constant(0.0, name='dp0')

    dp_offsets = np.linspace(-4e-3, +4e-3, n_div)
    dp0vec = dp0 + dp_offsets

    t0 = time.time()
    x1, dp1 = x0, dp0vec
    with tf.GradientTape(persistent=False) as tape:
        tape.watch(x0)
        tape.watch(dp0)

        for _ in range(n_repeat):
            x1 = simple_func(x1, dp1)
    print(f'Forward pass took {time.time()-t0:.3f} [s]')

    t0 = time.time()
    m11 = tape.jacobian(x1, x0)
    print(f'Gradient calc for m11 took {time.time()-t0:.3f} [s]')

    print('Finished')

    return m11

if __name__ == '__main__':

    n_repeat = 100
    n_div = 1001

    print(f'n_repeat = {n_repeat:d}, n_div = {n_div:d}')
    t0 = time.time()
    m11_list_vec = vector_example(n_repeat, n_div)
    tElapsed_vec = time.time()-t0

    t0 = time.time()
    dp_offsets = np.linspace(-4e-3, +4e-3, n_div, dtype=np.float32)
    m11_list_sca = np.array([scalar_example(n_repeat, dp0=dp0, show_print=False).numpy()
                             for dp0 in dp_offsets])
    tElapsed_sca = time.time()-t0

    print([n_repeat, n_div, m11_list_vec.numpy(),
           m11_list_sca,
           tElapsed_vec, tElapsed_sca])
```

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@yhidaka 
Please refer to the [gist here](https://colab.research.google.com/gist/Saduf2019/4275c32bc20bc39ec7feba8f98402302/untitled585.ipynb) for the performance on tf 2.4 and tf 2.5 rc0 and let us know.

**Comment by yhidaka** [comment]
@Saduf2019 
Thank you for your comment. I ran your gist with 3 and 1001 for "n_div" and got the similar results for "tElapsed_vec" (using tape.jacobian) and "tElapsed_sca" (using tape.gradient) with both tf2.4 & 2.5:

# tf2.4.1

n_div = 3
Trial-1: jac = 5.3 [s]; grad = 0.08 [s]
Trial-2: jac = 5.4 [s]; grad = 0.08 [s]

n_div = 1001
Trial-1: jac = 6.0 [s]; grad = 27.3 [s]
Trial-2: jac = 5.5 [s]; grad = 27.6 [s]

# tf2.5.0-rc0

n_div = 3
Trial-1: jac = 5.2 [s]; grad = 0.10 [s]
Trial-2: jac = 5.4 [s]; grad = 0.09 [s]

n_div = 1001
Trial-1: jac = 5.5 [s]; grad = 29.6 [s]
Trial-2: jac = 5.4 [s]; grad = 29.7 [s]

#

So, my question still remains. The vector case (using tape.jacobian) and scalar case (using tape.gradient) return exactly the same gradient array (both in terms of size [= "n_div"] and value), yet the first case takes nearly constant time (and apparently with significant overhead time for setup), while the computation time for the second case scales roughly proportionally to "n_div". Is this an expected behavior?

**Comment by kkimdev** [comment]
I'd say that's more-or-less expected for this smaller-scale computation, since constant Python code overhead dominates it.  Using `@tf.fucntion` could give a different picture here.
